slidechain/watch.go

```diff
@@ -117,62 +118,153 @@ func (c *Custodian) watchExports(ctx context.Context) {
-				}
+			c.exports.Broadcast()
+		}
+	}
+}
 
-				log.Printf("recorded export: %d of txvm asset %x (Stellar %x) for %s", retiredAmount, retiredAssetIDBytes, info.AssetXDR, exporter.Address())
+// Runs as a goroutine
+func (c *Custodian) watchPegOuts(ctx context.Context) {
+	defer log.Print("watchPegOuts exiting")
 
-				c.exports.Broadcast()
+	// Poll the database every minute for unprocessed exports.
+	go func(ctx context.Context) {
+		defer log.Print("export table poll exiting")
+		for t := time.Tick(1 * time.Minute); ; {
+			// Lock access to the export database.
+			// We do this to avoid double-processing any peg-out tx, via selecting it in both go-routines.
+			// We unlock at the end of an iteration of this go-routine or before any early return.
+			c.exportmux.Lock()
+			const q = `SELECT amount, asset_xdr, exporter, temp, seqnum, anchor, pubkey FROM exports WHERE (pegged_out=1 OR pegged_out=3)`
+			var (
+				txids, assetXDRs, anchors, pubkeys [][]byte
+				amounts, seqnums, peggedOuts       []int64
+				exporters, temps                   []string
+			)
+			err := sqlutil.ForQueryRows(ctx, c.DB, q, func(txid []byte, amount int64, assetXDR []byte, exporter, temp string, seqnum, peggedOut int64, anchor, pubkey []byte) {
+				txids = append(txids, txid)
+				amounts = append(amounts, amount)
+				assetXDRs = append(assetXDRs, assetXDR)
+				exporters = append(exporters, exporter)
+				temps = append(temps, temp)
+				seqnums = append(seqnums, seqnum)
+				peggedOuts = append(peggedOuts, peggedOut)
+				anchors = append(anchors, anchor)
+				pubkeys = append(pubkeys, pubkey)
+			})
+			if err == context.Canceled {
+				c.exportmux.Unlock()
+				return
+			}
+			if err != nil {
+				c.exportmux.Unlock()
+				log.Fatalf("querying peg-outs: %s", err)
+			}
+			for i, txid := range txids {
+				err = c.doPostExport(ctx, assetXDRs[i], anchors[i], txid, amounts[i], seqnums[i], peggedOuts[i], exporters[i], temps[i], pubkeys[i])
+				if err != nil {
+					c.exportmux.Unlock()
+					if err == context.Canceled {
+						return
+					}
+					log.Fatal(err)
+				}
+			}
+			c.exportmux.Unlock()
+			select {
+			case <-t:
+				continue
+			case <-ctx.Done():
+				return
+			}
+		}
+	}(ctx)
 
-				i++ // advance past the consumed log ("L") entry
+	for {
+		// The lock and unlock pattern for the export database is as in the above go routine.
+		c.exportmux.Lock()
+		// Read in new peg-outs via channel.
+		var pegOut PegOut
+		select {
+		case <-ctx.Done():
+			c.exportmux.Unlock()
+			return
+		case pegOut = <-c.pegouts:
+		}
+		const q = `SELECT amount, asset_xdr, exporter, temp, seqnum, anchor, pubkey FROM exports WHERE txid=$1 AND pegged_out=$2`
+		var (
+			assetXDR, anchor, pubkey []byte
+			amount, seqnum           int64
+			exporter, temp           string
+		)
+		err := sqlutil.ForQueryRows(ctx, c.DB, q, pegOut.txid, pegOut.state, func(qAmount int64, qAssetXDR []byte, qExporter, qTemp string, qSeqnum int64, qAnchor, qPubkey []byte) {
```

Comment:
Is it necessary to get this information from the database? Doesn't `pegOutFromExports` have it all, and can't it send it over the `c.pegouts` channel?

Also, it's never possible to have more than one row matching the given txid, so you want `c.DB.QueryRowContext` here, not `ForQueryRows`. (Although it is possible to have zero rows, which should perhaps be treated as an error.)